cdap-kms/src/main/java/co/cask/cdap/security/store/KMSSecureStore.java

```diff
@@ -0,0 +1,160 @@
+    options.setDescription(description);
+    options.setAttributes(properties);
+    options.setBitLength(data.length * Byte.SIZE);
+    String keyName = getKeyName(namespace, name);
+    try {
+      provider.createKey(keyName, data, options);
+    } catch (IOException e) {
+      throw new IOException("Failed to store the key. " + name + " under namespace " + namespace, e);
+    }
+  }
+
+  /**
+   * Deletes the element with the given name.
+   * @param namespace The namespace this key belongs to.
+   * @param name Name of the element to be deleted.
+   */
+  @Override
+  public void delete(String namespace, String name) throws IOException {
+    try {
+      provider.deleteKey(getKeyName(namespace, name));
+    } catch (IOException e) {
+      throw new IOException("Failed to delete the key. " + name + " under namespace " + namespace, e);
+    }
+  }
+
+  /**
+   * List of all the entries in the secure store.
+   * @return A list of {@link SecureStoreMetadata} objects representing the data stored in the store.
+   * @param namespace The namespace this key belongs to.
+   */
+  @Override
+  public List<SecureStoreMetadata> list(String namespace) throws IOException {
+    String prefix = namespace + NAME_SEPARATOR;
+    List<String> keysInNamespace = new ArrayList<>();
+    KeyProvider.Metadata[] metadatas;
+    try {
+      for (String key : provider.getKeys()) {
+        if (key.startsWith(prefix)) {
+          keysInNamespace.add(key);
+        }
+      }
+      metadatas = provider.getKeysMetadata(keysInNamespace.toArray(new String[0]));
+    } catch (IOException e) {
+      throw new IOException("Failed to get the list of elements from the secure store.", e);
+    }
+
+    List<SecureStoreMetadata> secureStoreMetadatas = new ArrayList<>();
+    for (int i = 0; i < metadatas.length; i++) {
+      KeyProvider.Metadata metadata = metadatas[i];
+      SecureStoreMetadata meta = SecureStoreMetadata.of(keysInNamespace.get(i).substring(prefix.length()),
+                                                        metadata.getDescription(),
+                                                        metadata.getAttributes());
+      secureStoreMetadatas.add(meta);
+    }
+    return secureStoreMetadatas;
+  }
+
+  /**
+   * @param namespace The namespace this key belongs to.
+   * @param name Name of the data element.
+   * @return An object representing the securely stored data associated with the name.
+   */
+  @Override
+  public SecureStoreData get(String namespace, String name) throws IOException {
+    String keyName = getKeyName(namespace, name);
+    KeyProvider.Metadata metadata = provider.getMetadata(keyName);
+    SecureStoreMetadata meta = SecureStoreMetadata.of(name, metadata.getDescription(), metadata.getAttributes());
+    KeyProvider.KeyVersion keyVersion = provider.getCurrentKey(keyName);
+    return new SecureStoreData(meta, keyVersion.getMaterial());
+  }
+
+  KeyProvider getProvider() {
```

Comment:
This is needed in the secure store Utils to get the delegation token.


> On Jul 19, 2016, at 10:58 AM, Terence Yim <notifications@github.com> wrote:
> 
> In cdap-kms/src/main/java/co/cask/cdap/security/store/KMSSecureStore.java <https://github.com/caskdata/cdap/pull/6066#discussion_r71390152>:
> 
> > +
> > +  /**
> > +   * @param namespace The namespace this key belongs to.
> > +   * @param name Name of the data element.
> > +   * @return An object representing the securely stored data associated with the name.
> > +   */
> > +  @Override
> > +  public SecureStoreData get(String namespace, String name) throws IOException {
> > +    String keyName = getKeyName(namespace, name);
> > +    KeyProvider.Metadata metadata = provider.getMetadata(keyName);
> > +    SecureStoreMetadata meta = SecureStoreMetadata.of(name, metadata.getDescription(), metadata.getAttributes());
> > +    KeyProvider.KeyVersion keyVersion = provider.getCurrentKey(keyName);
> > +    return new SecureStoreData(meta, keyVersion.getMaterial());
> > +  }
> > +
> > +  KeyProvider getProvider() {
> Shouldn't need to expose it.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/caskdata/cdap/pull/6066/files/e867ab05183313ef992bdfd396e440874e647424#r71390152>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAFGk0RU5EuuEwo133HskbXyGajWEUWKks5qXRBDgaJpZM4JOEs0>.
>